### PR TITLE
feat(event_handler): support Header parameter validation in OpenAPI schema

### LINKED ISSUE #3661: Feature request: Add support for headers parameters in OpenAPI schema
### Use case

Original discussion: https://github.com/aws-powertools/powertools-lambda-python/discussions/3609

OpenAPI supports Headers parameters and we must add support for this!

https://swagger.io/docs/specification/describing-parameters/

#### Header Parameters
An API call may require that custom headers be sent with an HTTP request. OpenAPI lets you define custom request headers as in: header parameters. 

### Solution/User Experience

Add `Header` in https://github.com/aws-powertools/powertools-lambda-python/blob/develop/aws_lambda_powertools/event_handler/openapi/params.py

### Alternative solutions

_No response_

### Acknowledgment

- [X] This feature request meets [Powertools for AWS Lambda (Python) Tenets](https://docs.powertools.aws.dev/lambda/python/latest/#tenets)
- [X] Should this be considered in other Powertools for AWS Lambda languages? i.e. [Java](https://github.com/aws-powertools/powertools-lambda-java/), [TypeScript](https://github.com/aws-powertools/powertools-lambda-typescript/), and [.NET](https://github.com/aws-powertools/powertools-lambda-dotnet/)

### PR DESCRIPTION
**Pull request by leandrodamascena**
**Issue number:** #3661 

### Changes

This pull request aims to enhance the OpenAPI schema by allowing the definition of headers as part of API specifications. An API call may require that custom headers be sent with an HTTP request. OpenAPI lets you define custom request headers as in: header parameters.

##### Key Changes:

- Added support for defining custom headers within the OpenAPI schema.
- Updated the OpenAPI documentation to showcase the usage of custom header parameters.
- Refactored the tests to reduce the number of the code

### User experience

```python
from typing import List

from aws_lambda_powertools.event_handler import APIGatewayRestResolver
from aws_lambda_powertools.event_handler.openapi.params import Header
from aws_lambda_powertools.shared.types import Annotated  
from aws_lambda_powertools.utilities.typing import LambdaContext

app = APIGatewayRestResolver(enable_validation=True)

@app.get("/hello")
def get_hello(header2: Annotated[List[str], Header()], header1: Annotated[str, Header()]):
    print(header2)

def lambda_handler(event: dict, context: LambdaContext) -> dict:
    return app.resolve(event, context)
```

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by leandrodamascena** [comment]
@rubenfonseca this PR is ready to review. I want to highlight one point that you may want to change/refactor.

According to the RFC, headers are case insensitive, so I lower the header names when resolving them in the DataClass and lower the parameter name as well when validating the OpenAPI schema. Even this is against Python standards, but a customer can create a function like this:

```python
@app.get("/hello")
def get(
    HeADer1: Annotated[
        str,
        Header(
            description="This is multi value header parameter.",
        ),
    ],
):
    """Return validated multi-value header values."""
    return HeADer1
```

We must be able to compare "HeADer1" with the header name "header1" or "Heder1" or "HEader1".

Thanks

**Review by rubenfonseca — CHANGES_REQUESTED** [review]
This is great work, I only need to check the documentation now.
As we spoke during peer review, try to see if we can move the `.lower` inside the Header's class alias setter. This way we can remove some if/elses in the code and streamline the feature.
Great work Leandro!

File: docs/core/event_handler/api_gateway.md [hunk 1]
```diff
@@ -424,6 +424,40 @@ For example, we could validate that `<todo_id>` dynamic path should be no greate
 
 1. `Path` is a special OpenAPI type that allows us to constrain todo_id to be less than 999.
 
+#### Validating headers
+
+We use the `Annotated` type to tell Event Handler that a particular parameter is a header that needs to be validated.
```
Comment (rubenfonseca) [inline docs/core/event_handler/api_gateway.md]:
```suggestion
We use the `Annotated` type to tell the Event Handler that a particular parameter is a header that needs to be validated.
```

File: docs/core/event_handler/api_gateway.md [hunk 2]
```diff
@@ -424,6 +424,40 @@ For example, we could validate that `<todo_id>` dynamic path should be no greate
...
+
+!!! info "We adhere to HTTP RFC standards, which means we treat HTTP headers as case-insensitive."
```
Comment (rubenfonseca) [inline docs/core/event_handler/api_gateway.md]: Can we add a link to the RFC standards like you did in the code?

File: docs/core/event_handler/api_gateway.md [hunk 3]
```diff
@@ -424,6 +424,40 @@ For example, we could validate that `<todo_id>` dynamic path should be no greate
...
+
+In the following example, we use a new `Header` OpenAPI type to add [one out of many possible constraints](#customizing-openapi-parameters), which should read as:
+
+* `correlation_id` is a header that must be present in the request
+* `correlation_id`, when set, should have 16 characters
+* Doesn't match? Event Handler will return a validation error response
```
Comment (rubenfonseca) [inline docs/core/event_handler/api_gateway.md]:
```suggestion
* No match? Event Handler will return a validation error response
```

File: docs/core/event_handler/api_gateway.md [hunk 4]
```diff
@@ -424,6 +424,40 @@ For example, we could validate that `<todo_id>` dynamic path should be no greate
...
+#### Validating headers
+
+We use the `Annotated` type to tell Event Handler that a particular parameter is a header that needs to be validated.
+
+!!! info "We adhere to HTTP RFC standards, which means we treat HTTP headers as case-insensitive."
+
+In the following example, we use a new `Header` OpenAPI type to add [one out of many possible constraints](#customizing-openapi-parameters), which should read as:
+
+* `correlation_id` is a header that must be present in the request
+* `correlation_id`, when set, should have 16 characters
```
Comment (rubenfonseca) [inline docs/core/event_handler/api_gateway.md]:
```suggestion
* `correlation_id` should have 16 characters
```

The previous bullet point already makes it required

File: docs/core/event_handler/api_gateway.md [hunk 5]
```diff
@@ -424,6 +424,40 @@ For example, we could validate that `<todo_id>` dynamic path should be no greate
...
+<!-- markdownlint-disable MD013 -->
+
+=== "validating_headers.py"
+
+    ```python hl_lines="8 10 29"
+    --8<-- "examples/event_handler_rest/src/validating_headers.py"
+    ```
+
+    1. If you're not using Python 3.9 or higher, you can install and use [`typing_extensions`](https://pypi.org/project/typing-extensions/){target="_blank" rel="nofollow"} to the same effect
+    2. `Header` is a special OpenAPI type that can add constraints to a header well as document them
```
Comment (rubenfonseca) [inline docs/core/event_handler/api_gateway.md]:
```suggestion
    2. `Header` is a special OpenAPI type that can add constraints and documentation to a header
```

File: docs/core/event_handler/api_gateway.md [hunk 6]
```diff
@@ -424,6 +424,40 @@ For example, we could validate that `<todo_id>` dynamic path should be no greate
...
+    3. **First time seeing the `Annotated`?** <br><br> This special type uses the first argument as the actual type, and subsequent arguments are metadata. <br><br> At runtime, static checkers will also see the first argument, but anyone receiving them could inspect them to fetch their metadata.
```
Comment (rubenfonseca) [inline docs/core/event_handler/api_gateway.md]:
```suggestion
    3. **First time seeing `Annotated`?** <br><br> This special type uses the first argument as the actual type, and subsequent arguments as metadata. <br><br> At runtime, static checkers will also see the first argument, but any receiver can inspect it to get the metadata.
```

File: docs/core/event_handler/api_gateway.md [hunk 7]
```diff
@@ -424,6 +424,40 @@ For example, we could validate that `<todo_id>` dynamic path should be no greate
...
+
+=== "working_with_headers_multi_value.py"
+
+    If you need to handle multi-value for specific headers, you can create a list of the desired type.
```
Comment (rubenfonseca) [inline docs/core/event_handler/api_gateway.md]:
```suggestion
    You can handle multi-value headers by declaring it as a list of the desired type.
```

**Review by rubenfonseca — CHANGES_REQUESTED** [review]
Great job with the code, much cleaner! Just left some final comments on the documentation.

**Review by rubenfonseca — APPROVED** [review]

**Pull request merged**